### DISCUSSION TITLE
How to use Drift isolates, drift v2.6 and flutter 3.10.5

### DISCUSSION THREAD

**Discussion by MaxiStefan** [simolus3/drift#2496] [Q&A]
Hi,

So far I've been passing my db file (String) to another isolate and opened the connection to the database over there for any heavy workload (i). However, with my sdk upgrade, I wanted to try the computeWithDatabase / serializableConnection methods to execute my workload on isolates and by following the example on https://drift.simonbinder.eu/docs/advanced-features/isolates/ I encounter the same non-primitive illegal argument exception that made me do things as mentioned in **(i)**. 

How does one make the example Future<void> customIsolateUsage(MyDatabase database) work when using the Isolate.run() method.

**Comment by simolus3** [comment]
> encounter the same non-primitive illegal argument exception

Recent Dart SDKs (which should be included in Flutter) should print a path to the illegal object being sent with relevant field names, right? Do you see that message or is this a generic error message?

**Comment by orestesgaolin** [comment]
This is a simplified implementation of how I deal with isolates with encrypted database. Not sure if this is 100% correct as it sometimes is prone to database locking. I wonder if there are any obvious errors that I made here @simolus3 ? Maybe something causing the database to be created multiple times and thus locking?

```dart
class LocalDatabaseClient {
  LocalDatabaseClient({
    required this.credentials,
    Logger? log,
  }) : _log = log ?? Logger(tag: 'LDC');

  LocalDatabase? _bgDatabase;
  final Logger _log;
  final LocalDatabaseCredentials credentials;

  /// Generic function to retry a query that might fail due to
  /// [SqliteException] e.g. when database is locked.
  Future<T> retry<T>(Future<T> Function() computation) async {
    var retryCount = 0;

    while (retryCount < 5) {
      try {
        return computation();
      } on SqliteException catch (e, s) {
        if (retryCount == 5) {
          Error.throwWithStackTrace(e, s);
        }
        if (retryCount == 4) {
          // Last chance to fix the access issue, let's close the connection
          // and try to reconnect
          _log.w('Closing database to recover from access issue');
          await _bgDatabase?.close();
          _bgDatabase = null;
        }
        retryCount++;
        _log.e('Problem with accessing the database, retry $retryCount', e);
      }
      await Future<void>.delayed(const Duration(milliseconds: 500));
    }
    _log.e('Did not manage to recover the database despite retrying');
    throw Exception('Did not manage to recover the database despite retrying');
  }

  Future<List<Data>> getData() async {
    _log.d('Getting data that need to be uploaded.');

    try {
      return retry(() async {
        final db = await _ensureBgDatabaseAvailable();

        final query = db.dataTable.select()
          ..where((tbl) => tbl.uploaded.equals(false))
          ..orderBy([
            (u) => OrderingTerm(
                  expression: u.timestamp,
                  // ignore: avoid_redundant_argument_values
                  mode: OrderingMode.asc,
                ),
          ]);

        final result = await query.get();

        _log.d(
          'There are ${result.length} that need to be uploaded.',
        );

        return result;
      });
    } catch (e, s) {
      _log.e('Getting data that need to be updated failed.', e, s);
      return [];
    }
  }

  Future<LocalDatabase> _ensureBgDatabaseAvailable() async {
    if (_bgDatabase == null) {
      final isolate = await createIsolateWithSpawn();
      _bgDatabase = LocalDatabase.connect(
        await isolate.connect(singleClientMode: true),
        credentials,
      );
      return _bgDatabase!;
    }
    return _bgDatabase!;
  }

  Future<DriftIsolate> createIsolateWithSpawn() async {
    final token = RootIsolateToken.instance!;
    return DriftIsolate.spawn(() {
      open.overrideFor(OperatingSystem.android, openCipherOnAndroid);
      BackgroundIsolateBinaryMessenger.ensureInitialized(token);

      // The callback to DriftIsolate.spawn() must return the database
      // connection to use.
      return _openDatabase(credentials: credentials);
    });
  }
}

@DriftDatabase(
  tables: [
    DataTable,
  ],
)
class LocalDatabase extends _$LocalDatabase {
  LocalDatabase({
    required this.credentials,
  }) : super(_openDatabase(credentials: credentials)) {
    // https://github.com/simolus3/drift/blob/develop/examples/encryption/lib/main.dart#L11
    open.overrideFor(OperatingSystem.android, openCipherOnAndroid);

    /// Since we open one database per user, we don't use the database as a
    /// singleton. Hence, we disable the warning thrown by the library.
    driftRuntimeOptions.dontWarnAboutMultipleDatabases = true;
  }

  LocalDatabase.connect(super.e, this.credentials) {
    // connection.runCustom('pragma journal_mode=WAL;');
    driftRuntimeOptions.dontWarnAboutMultipleDatabases = true;
  }

  final LocalDatabaseCredentials credentials;

  /// Schema version has to match with the native database.
  /// - [Android](packages/bluetooth_foreground_service/android/src/main/kotlin/com/example/bluetooth_foreground_service/LocalDatabaseClient.kt)
  /// - [iOS](packages/bluetooth_foreground_service/ios/Classes/LocalDatabaseClient.swift)
  @override
  int get schemaVersion => 3;

  /// Migrations are handled natively.
  /// - [Android](packages/bluetooth_foreground_service/android/src/main/kotlin/com/example/bluetooth_foreground_service/LocalDatabaseClient.kt)
  /// - [iOS](packages/bluetooth_foreground_service/ios/Classes/LocalDatabaseClient.swift)
  @override
  MigrationStrategy get migration => MigrationStrategy(
        beforeOpen: (details) async {
          await customStatement('pragma journal_mode=WAL;');
          await customStatement('PRAGMA foreign_keys = ON');
          await customStatement('pragma cipher_version');
        },
        onUpgrade: (m, from, to) async {
          Logger(tag: 'LDC').i(
            'Not migrating from $from to $to as it should be handled natively',
          );
        },
      );
}

QueryExecutor _openDatabase({required LocalDatabaseCredentials credentials}) {
  return LazyDatabase(() async {
    late final Directory dbDirectory;

    /// On Android, the `getDatabasesPath` utility function is used to obtain
    /// the database path. It will return `data/data/<package_name>/databases`.
    ///
    /// On iOS, as suggested by the documentation of `getDatabasePath`,
    /// the `path_provider` is recommended to get the databases directory.
    ///
    /// The most appropriate location on iOS would be the Library directory that
    /// you could get from the `path_provider.getLibraryDirectory` function.
    ///
    /// The `getLibraryDirectory` provides a path to the directory where
    /// application can store files that are persistent, backed up, and not
    /// visible to the user, such as sqlite.db.
    if (Platform.isAndroid) {
      dbDirectory = Directory(await getDatabasesPath());
    } else if (Platform.isIOS) {
      dbDirectory = await getLibraryDirectory();
    } else {
      throw Exception('Unsupported platform.');
    }

    final dbPath = p.join(
      dbDirectory.path,
      credentials.databaseName,
    );

    return NativeDatabase(
      File(dbPath),
      setup: (db) {
        try {
          // Check that we're actually running with SQLCipher by querying the
          // cipher_version pragma.
          final result = db.select('pragma cipher_version');
          if (result.isEmpty) {
            throw UnsupportedError(
              'This database needs to run with SQLCipher, but that library is '
              'not available!',
            );
          }

          db
            // Set the database password.
            ..execute("pragma key = '${credentials.databasePass}'")
            ..execute('pragma journal_mode=WAL;')
            ..execute('pragma busy_timeout = 100;')

            // Test that the key is correct by selecting from a table.
            ..execute('select count(*) from sqlite_master');
        } catch (e, s) {
          Error.throwWithStackTrace(e, s);
        }
      },
    );
  });
}
```

**Reply by simolus3** [reply]
The opening logic looks reasonable. There's a race in `_ensureBgDatabaseAvailable` if it's called in quick succession before the awaits complete and `_bgDatabase` is assigned. You have disabled the warning for multiple database instances on a single isolate, could that be related?
And it doesn't hurt, but I've noticed you have some statements both in `beforeOpen` and in `setup`. `setup` gets executed first, so that's a good place to enable WAL. You don't need to do that in the migration callback as well (moving the foreign key pragma might change semantics if you have weird migrations, but it could be done too for consistency).

Does the retry loop improve things for you? Setting a busy timeout will essentially make sqlite3 do the same thing internally, so I'm not sure if another wrapper in Dart improves the situation.
And you mention that the migration is handled natively - if you're opening the database in native code, are you also applying WAL mode and a busy timeout there?

Do you know if you have any long-running transactions? (More than the 100ms?). Such transactions on one isolate could block the database in others. It's unlikely that writes alone take so long, but this is something that could happen if you make a network call in a `transactiom` block for instance (probably something you don't want to do).

**Reply by orestesgaolin** [reply]
> There's a race in _ensureBgDatabaseAvailable

I thought about adding additional completer there just in case this method is called in succession. Maybe a good shout!

> And it doesn't hurt, but I've noticed you have some statements both in beforeOpen and in setup

I was wondering which is _better_ place to put these pragma calls? If you reckon that `setup` is better then maybe I'll move all the pragma calls there?

> Does the retry loop improve things for you?

Hard to tell, but the main reason was mostly to "directly" capture SqliteExceptions instead of generalized `catch` in the upper layer of the app.

> And you mention that the migration is handled natively 

In this particular case the database is never migrated, the comment is leftover from a bit more complex database (which is not written to from Flutter side). 

> if you're opening the database in native code, are you also applying WAL mode and a busy timeout there?

Yes, I'm applying WAL there, but I forgot about busy timeout, I'll add it there as well.

> Do you know if you have any long-running transactions?

Not sure about the timing, but I've been able to reproduce database locking when trying to write at the same time from main Flutter isolate on app launch and from a background task (using [workmanager](https://pub.dev/packages/workmanager) plugin). The background task was meant to be run overnight but it seems that on some devices it's getting executed only when the app gets launched (maybe due to power management rules of the operating system?) so they kick-off at the same time and from time to time it can lead to concurrent writes. Wondering if you have any good workaround for this kind of issues?

**Reply by simolus3** [reply]
> If you reckon that setup is better then maybe I'll move all the pragma calls there?

Conceptually, `setup` runs before drift does anything with the database. `beforeOpen` runs before the database is considered fully opened and usable, but _after_ your migrations. Setting the key, the WAL mode and the busy timeout should be done before any write, so these belong into `setup`. I think you can do the same with the foreign key pragma, but that is also fine in the migration callback. If you prefer to have them all in one place, move them into `setup`.

> Hard to tell, but the main reason was mostly to "directly" capture SqliteExceptions instead of generalized catch in the upper layer of the app.

Note that, since you're using drift isolates, you'll likely get a `DriftRemoteException` instead of a raw `SqliteException`. We might change this in a future breaking release of drift, it's an artifact of not being able to send arbitrary objects (like those exceptions) across isolates for a long time.

> so they kick-off at the same time and from time to time it can lead to concurrent writes. Wondering if you have any good workaround for this kind of issues?

So the thing is that, with a `busy_timeout`, concurrent writes should not really get in each other's way. A write will lock the database (for a very short time), and the other writer will briefly wait and retry afterwards if the busy timeout is set.

I've tried to reproduce this by starting two isolates that open the same database and write to it in an infinite loop. Without the busy timeout, I get the locked errors very quickly. So far, I've not been able to reproduce the issue with timeouts enabled.

**Reply by orestesgaolin** [reply]
Maybe it will be interesting to you, but with busy_timeout of 100 the writes still can fail, even when there's no other process writing at the same time.

The logs below are from the background process run via the workmanager plugin during the night when user was not actively using the app


```
2023-07-19T00:46:28.916715Z SynchronizationRepository: Starting sync, mode SyncMode.background, id: 37
2023-07-19T00:46:29.471385Z DataClient: Auth state changed
2023-07-19T00:46:29.882840Z DataClient: Got the user from backend
2023-07-19T00:46:31.124897Z SynchronizationRepository: Got [redacted]
2023-07-19T00:46:31.125256Z VDBC: Saving [redacted]
2023-07-19T00:46:31.133560Z SynchronizationRepository: Got app configuration
2023-07-19T00:46:31.134142Z VDBC: Saving [redacted]: 37
2023-07-19T00:46:31.141266Z SynchronizationRepository: Got 5 [redacted]
2023-07-19T00:46:31.141339Z VDBC: Saving [redacted]: 5
2023-07-19T00:46:31.146452Z SynchronizationRepository: Saved [redacted] to database
2023-07-19T00:46:31.146493Z SynchronizationRepository: Getting [redacted]
2023-07-19T00:46:31.146576Z VDBC: Saving [redacted]
2023-07-19T00:46:31.151643Z SynchronizationRepository: Saved [redacted]
2023-07-19T00:46:31.151683Z SynchronizationRepository: Got user [redacted]
2023-07-19T00:46:31.151708Z VDBC: Saving [redacted]
2023-07-19T00:46:31.159826Z SynchronizationRepository: SqliteException(5): while executing statement, database is locked, database is locked (code 5)
2023-07-19T00:46:31.159855Z SynchronizationRepository:   Causing statement: UPDATE "user_[redacted]_table" SET "[redacted]" = ?, "[redacted]" = ?, "other_[redacted]" = ?, "updated_at" = ? WHERE "id" = ?;, parameters: 1, 0, null, 1669131117, 0
2023-07-19T00:46:31.159872Z SynchronizationRepository: #0   DriftCommunication.request (package:drift/src/remote/communication.dart:113)
2023-07-19T00:46:31.159885Z SynchronizationRepository: #1   _BaseExecutor._runRequest (package:drift/src/remote/client_impl.dart:97)
2023-07-19T00:46:31.159898Z SynchronizationRepository: #2   _BaseExecutor.runDelete (package:drift/src/remote/client_impl.dart:114)
2023-07-19T00:46:31.159912Z SynchronizationRepository: #3   UpdateStatement._performQuery.<anonymous closure> (package:drift/src/runtime/query_builder/statements/update.dart:37)
2023-07-19T00:46:31.159926Z SynchronizationRepository: #4   DatabaseConnectionUser.doWhenOpened.<anonymous closure> (package:drift/src/runtime/api/connection_user.dart:162)
2023-07-19T00:46:31.159940Z SynchronizationRepository: #5   _rootRunUnary (dart:async/zone.dart:1407)
2023-07-19T00:46:31.159952Z SynchronizationRepository: #6   <asynchronous suspension>
2023-07-19T00:46:31.159964Z SynchronizationRepository: #7   UpdateStatement._performQuery (package:drift/src/runtime/query_builder/statements/update.dart:36)
2023-07-19T00:46:31.159977Z SynchronizationRepository: Synchronization failed for task 37
```

Here's the implementation of few writes:

The one that not failed before 2023-07-19T00:46:31.151708Z (146452Z - 141266Z = 5 ms)

```dart
  Future<void> saveRedacted(List<RedactedEntry> redacted) async {
    log.d('Saving Redacted: ${redacted.length}'); // 2023-07-19T00:46:31.141339Z
    final db = await ensureBgDatabaseAvailable();
    await db.batch((batch) {
      batch.insertAll(
        db.RedactedTable,
        redacted,
        mode: InsertMode.insertOrReplace,
      );
    });
  }
```

The one that failed at 2023-07-19T00:46:31.151708Z using a transaction

```dart
  Future<void> saveRedacted(UserRedactedEntry redacted) async {
    log.d('Saving redacted'); // 2023-07-19T00:46:31.151708Z
    final db = await ensureBgDatabaseAvailable();
    await db.transaction(() async {
      final currentRedacted =
          await db.select(db.userRedactedTable).getSingleOrNull();
      if (currentRedacted == null) {
        return db.into(db.userredactedTable).insert(
              redacted,
              mode: InsertMode.insertOrReplace,
            );
      } else {
        return db.update(db.userRedactedTable).replace(redacted);
      }
    });
  }
```


The above functions were written in different times and thus the difference in the implementation, but it's interesting to see that one of them failed with the database locked exception while the other was fine.

**Reply by simolus3** [reply]
> The logs below are from the background process run via the workmanager plugin during the night when user was not actively using the app

Hm, that's weird. I don't think the way these functions are written should make a difference like that. But if the database is locked, there [_has_ to be a second connection somewhere](https://sqlite.org/rescode.html#busy) (we'd get result code 6 if it was something in the same connection). Could you add a log call in `_openDatabase` that prints `Isolate.current.controlPort.hashCode` (to uniquely identify the isolate in a process run) and the path being opened?  Perhaps there is a case where, in a background tasks, two databases (with two drift isolates) are being opened by accident?
